File: backend/app/schemas/sleep.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.sleep_log import WakeFeeling


class SleepLogCreate(BaseModel):
    sleep_at: datetime
    wake_at: datetime
    quality: int = Field(ge=1, le=5)
    wake_feeling: WakeFeeling
    notes: str | None = Field(default=None, max_length=500)
    idempotency_key: str | None = None

    @model_validator(mode="after")
    def wake_after_sleep(self) -> "SleepLogCreate":
        if self.wake_at <= self.sleep_at:
            raise ValueError("Horário de acordar precisa ser depois do horário de dormir")
        return self


class SleepLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sleep_at: datetime
    wake_at: datetime
    quality: int
    wake_feeling: WakeFeeling
    notes: str | None
    duration_minutes: int
    # Dia de calendário (fuso do usuário) a que a noite pertence — o dia em
    # que a pessoa ACORDOU, não o dia em que deitou. Ver _serialize em
    # routers/sleep.py.
    log_date: date

File: mobile/src/api/sleep.ts
import { api } from "./client";

export type WakeFeeling = "descansado" | "cansado" | "muito_cansado";

export type SleepLog = {
  id: number;
  sleep_at: string;
  wake_at: string;
  quality: number;
  wake_feeling: WakeFeeling;
  notes: string | null;
  duration_minutes: number;
  /** Dia (YYYY-MM-DD) a que a noite pertence, já calculado no fuso do
   * usuário pelo backend — é o dia em que ele ACORDOU, não em que deitou. */
  log_date: string;
};

export async function listSleepLogs(): Promise<SleepLog[]> {
  const { data } = await api.get<SleepLog[]>("/sleep");
  return data;
}

/** Mesmo mecanismo de `logWeight`/`logMeal` — ver api/weight.ts. */
function chaveDeRegistro(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
}

export async function logSleep(payload: {
  sleep_at: string;
  wake_at: string;
  quality: number;
  wake_feeling: WakeFeeling;
  notes?: string | null;
}): Promise<SleepLog> {
  const { data } = await api.post<SleepLog>("/sleep", {
    ...payload,
    idempotency_key: chaveDeRegistro(),
  });
  return data;
}

export async function deleteSleepLog(id: number): Promise<void> {
  await api.delete(`/sleep/${id}`);
}
